MekHQ/src/mekhq/campaign/Campaign.java

```diff
@@ -7853,25 +7854,52 @@ public long getForceValue(boolean noInfantry) {
             if (noInfantry && u.getEntity() instanceof Infantry && !(u.getEntity() instanceof BattleArmor)) {
             	continue;
             }
-            // lets exclude dropships and jumpships
-            if (u.getEntity() instanceof Dropship
-                || u.getEntity() instanceof Jumpship) {
-                continue;
-            }
-            // we will assume sale value for now, but make this customizable
-            if (getCampaignOptions().useEquipmentContractSaleValue()) {
-                value += u.getSellValue();
+            if (u.getEntity() instanceof Dropship) {
```

Comment:
The logic seems sound, but I'd suggest creating a function that encapsulates the interior of the individual if blocks so we don't have almost identical logic copy and pasted repeatedly.

Additionally, I believe we're trying to move towards checking ETYPE_whatever flags rather than instanceof checks.

Reply:
ETYPEs are done but still not sure how to make any changes to the if statement. Each block calls of a different function so adding a new function to process it just moves the bulk of the code and doesn't really do much else. Did you have any ideas in mind?

Reply:
Sometimes that is appropriate to make methods smaller.

Reply:
Basically, I was suggesting that you take lines 7859 - 7866 and move them to a function that takes the results of get[whatever]ContractPercent as an int and returns the computed value. The only difference between the four clauses in that if/else block is which "get[whatever]ContractPercent" method you call, so you'll reduce the code volume by a factor of about 4.

Reply:
@NickAragua  @Dylan-M 
Ok, so before I push this, this is what I came up with. If you are fine with it I will go ahead and push the changes.
`public long getEquipmentContractValue(Unit u, long etype, boolean useSaleValue) {
        long value;
        long percentValue = 0;
        if (useSaleValue) {
            value = u.getSellValue();
        } else {
            value = u.getBuyCost();
        }

        if (etype == Entity.ETYPE_DROPSHIP) {
            percentValue = (long) ((getCampaignOptions().getDropshipContractPercent() / 100) * value);
        } else if (etype == Entity.ETYPE_WARSHIP) {
            percentValue = (long) ((getCampaignOptions().getWarshipContractPercent() / 100) * value);
        } else if (etype == Entity.ETYPE_JUMPSHIP) {
            percentValue = (long) ((getCampaignOptions().getJumpshipContractPercent() / 100) * value);
        } else {
            percentValue = (long) ((getCampaignOptions().getEquipmentContractPercent() / 100) * value);
        }

        return percentValue;
    }`

This would remove the contract percent if/else statements from the forcevalue function.

Reply:
not sure why the code comment is split like that :(

Reply:
That's an equally valid approach to what I had in mind. You probably don't need the casts to long, but thumbs up otherwise.

Reply:
That's not going to work. The etype is a bitmask of every flag in the class hierarchy, so for jumpships `getEntityType()` returns (`Entity.TYPE_AERO | Entity.TYPE_JUMPSHIP`), warships return (`Entity.TYPE_AERO | Entity.TYPE_JUMPSHIP | Entity.TYPE_WARSHIP`), and dropships return (`Entity.TYPE_AERO | Entity.TYPE_SMALLCRAFT | Entity.Type.DROPSHIP`). `getEntityType() == Entity.TYPE_JUMPSHIP, et. al. will always return false.

Some time ago I added a utility function to `Entity` to make etype checks more readable: hasETypeFlag(long).

Reply:
@pheonixstorm If you can get this cleaned up with the comments above from @neoancient I'll get this reviewed and pulled in.

Reply:
Had hoped to get it done last weekend but ended up getting really busy this week. Will try to get it done soon, hopefully today if all goes well.

Reply:
Using the new hasEType function makes it easier to parse. I think it will make removing all of the unit related instanceof from the code very easy. Anyway, new code pushed so let me know if it needs further changes.